src/RTTI/RTTIUtil.hpp

```diff
@@ -0,0 +1,144 @@
+#include <Reflection/BsRTTIType.h>
+
+/**
+ * \file
+ *
+ * Contains utility makros to have less noise inside the actual components for RTTI.
+ * RTTI needs at least two things to be working:
+ *
+ *  1. RTTI-class which lists the fields to serialize and has methods to
+ *     create a new instance of the class it describes.
+ *  2. in the actual component to get an instance of the RTTI-class.
+ *
+ * The second part is cluttering the component definition and is always the same,
+ * hence a makro is useful here.
+ */
+
+/**
+ * For use in the actual components header. Declares the functions for accessing
+ * the RTTI-instance of the component.
+ *
+ * Example:
+ *
+ *    class MyComponent : public bs::Component
+ *    {
+ *    public:
+ *      REGOTH_COMPONENT_DECLARE_RTTI(MyComponent)
+ *
+ *    protected:
+ *      MyComponent() = default; // For RTTI
+ *    }
+ *
+ * The parameterless constructor is still required to be defined manually. Also
+ * note that there is not semicolon after the makro!
+ * Futhermore, the makro invocation needs to be in `public`-land of the component.
+ *
+ * @note There seems to be a problem with `friend` declaring the RTTI-class
+ *       since that doesn't seem to get access to the constructor. So you
+ *       might have to make the parameterless constructor public until that
+ *       is figured out.
+ */
+#define REGOTH_DECLARE_RTTI(classname)      \
+  friend class RTTI_##classname;            \
+  static bs::RTTITypeBase* getRTTIStatic(); \
+  bs::RTTITypeBase* getRTTI() const override;
+
+/**
+ * For use in the actual components source. Defines the functions for accessing
+ * the RTTI-instance of the component.
+ *
+ * Example:
+ *
+ *     namespace REGoth
+ *     {
+ *       REGOTH_COMPONENT_DEFINE_RTTI(MyComponent)
+ *     }
+ *
+ * The current namespace needs direct access to the components class.
+ * Also note that there is not semicolon after the makro!
+ */
+#define REGOTH_DEFINE_RTTI(classname)          \
+                                               \
+  bs::RTTITypeBase* classname::getRTTIStatic() \
+  {                                            \
+    return RTTI_##classname::instance();       \
+  }                                            \
+                                               \
+  bs::RTTITypeBase* classname::getRTTI() const \
+  {                                            \
+    return classname::getRTTIStatic();         \
+  }
+
+/**
+ * For use in the RTTI-class. Implements the glue code to create an instance of the component
+ * the RTTI describes.
+ *
+ * Use this if the RTTI describes a class derived from bs::Component.
+ */
+#define REGOTH_IMPLEMENT_RTTI_CLASS_FOR_COMPONENT(classname)  \
+                                                              \
+  bs::SPtr<bs::IReflectable> newRTTIObject() override         \
+  {                                                           \
+    return bs::GameObjectRTTI::createGameObject<classname>(); \
```

Comment:
Could this be the reason why making the default constructors protected does not work, as the static method of GameObjectRTTI tries to access the protected constructor and not the RTTI class itself?

Reply:
In the newer bs::f version they seem to use `SceneObject::createEmptyComponent()` that's why they declare SceneObject as friend in the component classes.

Reply:
Accessing private and protected members from static functions should work fine, though. I haven't used friend classes much however...

Reply:
Nope, protected and private members can only be accessed if you explicitely grant access to them. Static functions are not exception to this.

Reply:
How do you explicitly grant access to them? Is that only a thing for friend-classes, because it works fine from within classes: https://godbolt.org/z/f5_2Z9

Ah, and now I understood what you meant! `createGameObject` is a template accessing the constructor, so the compiler *could* decide that `GameObjecrRTTI` is not a friend and has no access to it. Took me a minute, sorry :smile:

Reply:
Yes, exactly! And in the current master it seems like they moved the static function for creating empty components to the `SceneObject` class and therefore they made `SceneObject` friend to all components.

Reply:
I'll update bsf to fix this for good.